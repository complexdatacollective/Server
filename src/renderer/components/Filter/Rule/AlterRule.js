import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import { compose } from 'recompose';
import PropTypes from 'prop-types';
import cx from 'classnames';
import {
  has, includes, map, get,
} from 'lodash';
import { SortableElement } from 'react-sortable-hoc';
import DragHandle from './DragHandle';
import DropDown from './DropDown';
import Input from './Input';
// import { getCodebook } from '../../../selectors/protocol';
import { getVariableOptions } from './selectors';
import { getOperatorsForType } from './operators';

class AlterRule extends PureComponent {
  static propTypes = {
    id: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
    ]).isRequired,
    onUpdateRule: PropTypes.func,
    onDeleteRule: PropTypes.func,
    options: PropTypes.shape({
      type: PropTypes.string,
      operator: PropTypes.string,
      attribute: PropTypes.string,
      value: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.number,
      ]),
    }),
    nodeTypes: PropTypes.array,
    nodeAttributes: PropTypes.object,
    valueInputType: PropTypes.string,
    className: PropTypes.string,
  };

  static defaultProps = {
    options: {
      type: '',
      operator: '',
      attribute: '',
      value: '',
    },
    onUpdateRule: () => {},
    onDeleteRule: () => {},
    nodeTypes: [],
    nodeAttributes: {},
    valueInputType: null,
    className: '',
  };

  showAttributes() {
    return has(this.props.nodeAttributes, this.props.options.type);
  }

  showOperator() {
    return !!this.props.options.attribute;
  }

  showValue() {
    return !!this.props.options.operator
      && !includes(['EXISTS', 'NOT_EXISTS'], this.props.options.operator);
  }

  render() {
    const {
      id,
      nodeTypes,
      nodeAttributes,
      onUpdateRule,
      onDeleteRule,
      options: {
        type, operator, attribute, value,
      },
      valueInputType,
      className,
    } = this.props;

    return (
      <div className={cx('rule', 'rule--alter', className)}>
        <DragHandle />
        <div className="rule__options">
          <div className="rule__option rule__option--type">
            <DropDown
              options={nodeTypes}
              value={type}
              placeholder="{type}"
              onChange={(newValue) => onUpdateRule(newValue, id, 'type')}
            />
          </div>
          {this.showAttributes() && (
            <div className="rule__option rule__option--attribute">
              <DropDown
                options={has(nodeAttributes, type) ? nodeAttributes[type] : []}
                value={attribute}
                placeholder="{variable}"
                onChange={(newValue) => onUpdateRule(newValue, id, 'attribute')}
              />
            </div>
          )}
          {this.showOperator() && (
            <div className="rule__option rule__option--operator">
              <DropDown
                options={getOperatorsForType(valueInputType)}
                value={operator}
                placeholder="{rule}"
                onChange={(newValue) => onUpdateRule(newValue, id, 'operator')}
              />
            </div>
          )}
          {this.showValue() && (
            <div className="rule__option rule__option--value">
              <Input
                value={value}
                type={valueInputType}
                onChange={(newValue) => onUpdateRule(newValue, id, 'value')}
              />
            </div>
          )}
        </div>
        <button type="button" className="rule__delete" onClick={() => onDeleteRule(id)} />
      </div>
    );
  }
}

function mapStateToProps(state, { options, codebook }) {
  // const codebook = getCodebook(state);
  const nodeTypes = map(codebook.node, (node, nodeId) => [nodeId, node.name]);
  const valueInputType = options
    ? get(codebook.node, [options.type, 'variables', options.attribute, 'type'])
    : undefined;

  return {
    nodeTypes,
    nodeAttributes: getVariableOptions(codebook.node),
    valueInputType,
  };
}

export { AlterRule };

export default compose(
  SortableElement,
  connect(mapStateToProps),
)(AlterRule);
