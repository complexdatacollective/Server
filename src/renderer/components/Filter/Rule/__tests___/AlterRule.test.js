/* eslint-env jest */

import React from 'react';
import { shallow } from 'enzyme';
import { AlterRule } from '../AlterRule';

const mockProps = {
  id: '123',
  onUpdateRule: () => {},
  onDeleteRule: () => {},
  options: {
    type: '',
    operator: '',
    attribute: '',
    value: '',
  },
  nodeTypes: ['foo', 'bar'],
  valueInputType: undefined,
  nodeAttributes: {
    foo: ['baz', 'buzz'],
  },
};

xdescribe('<AlterRule />', () => {
  it('can render', () => {
    const component = shallow(<AlterRule {...mockProps} />);

    expect(component).toMatchSnapshot();
  });
});
