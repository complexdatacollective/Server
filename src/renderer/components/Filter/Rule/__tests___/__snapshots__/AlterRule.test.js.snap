// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AlterRule /> can render 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AlterRule
    className=""
    id="123"
    nodeAttributes={
      Object {
        "foo": Array [
          "baz",
          "buzz",
        ],
      }
    }
    nodeTypes={
      Array [
        "foo",
        "bar",
      ]
    }
    onDeleteRule={[Function]}
    onUpdateRule={[Function]}
    options={
      Object {
        "attribute": "",
        "operator": "",
        "type": "",
        "value": "",
      }
    }
    valueInputType={null}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <sortableHandle(DragHandle) />,
        <div
          className="rule__options"
        >
          <div
            className="rule__option rule__option--type"
          >
            <DropDown
              className={null}
              onChange={[Function]}
              options={
                Array [
                  "foo",
                  "bar",
                ]
              }
              placeholder="{type}"
              value=""
            />
          </div>
        </div>,
        <div
          className="rule__delete"
          onClick={[Function]}
        />,
      ],
      "className": "rule rule--alter",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="rule__option rule__option--type"
            >
              <DropDown
                className={null}
                onChange={[Function]}
                options={
                  Array [
                    "foo",
                    "bar",
                  ]
                }
                placeholder="{type}"
                value=""
              />
            </div>,
            false,
            false,
            false,
          ],
          "className": "rule__options",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <DropDown
                className={null}
                onChange={[Function]}
                options={
                  Array [
                    "foo",
                    "bar",
                  ]
                }
                placeholder="{type}"
                value=""
              />,
              "className": "rule__option rule__option--type",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "className": null,
                "onChange": [Function],
                "options": Array [
                  "foo",
                  "bar",
                ],
                "placeholder": "{type}",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": "div",
          },
          false,
          false,
          false,
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "rule__delete",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <sortableHandle(DragHandle) />,
          <div
            className="rule__options"
          >
            <div
              className="rule__option rule__option--type"
            >
              <DropDown
                className={null}
                onChange={[Function]}
                options={
                  Array [
                    "foo",
                    "bar",
                  ]
                }
                placeholder="{type}"
                value=""
              />
            </div>
          </div>,
          <div
            className="rule__delete"
            onClick={[Function]}
          />,
        ],
        "className": "rule rule--alter",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="rule__option rule__option--type"
              >
                <DropDown
                  className={null}
                  onChange={[Function]}
                  options={
                    Array [
                      "foo",
                      "bar",
                    ]
                  }
                  placeholder="{type}"
                  value=""
                />
              </div>,
              false,
              false,
              false,
            ],
            "className": "rule__options",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <DropDown
                  className={null}
                  onChange={[Function]}
                  options={
                    Array [
                      "foo",
                      "bar",
                    ]
                  }
                  placeholder="{type}"
                  value=""
                />,
                "className": "rule__option rule__option--type",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "className": null,
                  "onChange": [Function],
                  "options": Array [
                    "foo",
                    "bar",
                  ],
                  "placeholder": "{type}",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": "div",
            },
            false,
            false,
            false,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "rule__delete",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
